app/questionnaire/router.py

```diff
@@ -258,6 +268,140 @@ def _get_return_to_location_url(
                 "questionnaire.submit_questionnaire", _anchor=return_to_answer_id
             )
 
+    def _get_return_to_for_grand_calculated_summary(
+        self,
+        *,
+        return_to: str | None,
+        return_to_block_id: str | None,
+        location: Location,
+        routing_path: RoutingPath,
+        is_for_previous: bool,
+        return_to_answer_id: str | None = None,
+    ) -> str | None:
+        """
+        Builds the return url for a grand calculated summary,
+        and accounts for it possibly being in a different section to the calculated summaries it references
+        """
+        if not (return_to_block_id and self._schema.is_block_valid(return_to_block_id)):
+            return None
+
+        # Type ignore: if the block is valid, then we'll be able to find a section for it
+        grand_calculated_summary_section: str = self._schema.get_section_id_for_block_id(return_to_block_id)  # type: ignore
+        if grand_calculated_summary_section != location.section_id:
+            # the grand calculated summary is in a different section which will have a different routing path
+            # but don't go to it unless the current section is complete
+            if not self._progress_store.is_section_complete(location.section_id):
+                return self._get_return_url_for_inaccessible_location(
+                    is_for_previous=is_for_previous,
+                    return_to_block_id=return_to_block_id,
+                    return_to=return_to,
+                    routing_path=routing_path,
+                )
+
+            routing_path = self._path_finder.routing_path(
+                section_id=grand_calculated_summary_section
+            )
+        if self.can_access_location(
+            # grand calculated summaries do not yet support repeating sections, when they do, this will need to make use of list item id as well
+            Location(
+                block_id=return_to_block_id,
+                section_id=grand_calculated_summary_section,
+            ),
+            routing_path,
+        ):
+            return url_for(
+                "questionnaire.block",
+                block_id=return_to_block_id,
+                return_to=return_to,
+                _anchor=return_to_answer_id,
+            )
+        return self._get_return_url_for_inaccessible_location(
+            is_for_previous=is_for_previous,
+            return_to_block_id=return_to_block_id,
+            return_to=return_to,
+            routing_path=routing_path,
+        )
```

Comment:
Im not sure that we got this quite right, for the overlapping answers scenario, if you change the answer from the second CS then click Continue on first CS, then if you click Previous, you go to the Hub which doesn't seem right. We should stay in the same section if we can't route out?. If prev is GCS but we can't access it, go the block before it?